cli/sessionstore/sessionstore.go

```
@@ -0,0 +1,166 @@
+}
+
+const (
+	servicePrefix  = "coder-cli:"
+	keyringAccount = "session"
+)
+
+// Location represents where a session token is stored.
+type Location string
+
+const (
+	LocationNone    Location = "none"
+	LocationFile    Location = "file"
+	LocationKeyring Location = "keyring"
+)
+
+func serviceName(u *url.URL) string {
+	if u == nil || u.Host == "" {
+		return servicePrefix + "default"
+	}
+	host := strings.TrimSpace(strings.ToLower(u.Host))
+	return servicePrefix + host
+}
+
+// KeyringWithFallback is a Backend that prefers keyring storage and falls back to file
+// storage if the operating system keyring is unavailable. Happy path usage of this
+// type should start with NewKeyringWithFileFallback.
+type KeyringWithFallback struct {
+	keyringProvider keyringProvider
+}
+
+func NewKeyringWithFileFallback() KeyringWithFallback {
+	return NewKeyringProviderWithFileFallback(operatingSystemKeyring{})
+}
+
+func NewKeyringProviderWithFileFallback(provider keyringProvider) KeyringWithFallback {
+	return KeyringWithFallback{keyringProvider: provider}
+}
+
+// Read prefers reading the token from the operating system keyring and falls back
+// to file storage if the operating system keyring is unavailable.
+func (o KeyringWithFallback) Read(conf config.Root, serverURL *url.URL) (string, Location, error) {
+	svc := serviceName(serverURL)
+	tok, err := o.keyringProvider.Get(svc, keyringAccount)
+	if err == nil && tok != "" {
+		return tok, LocationKeyring, nil
+	}
+	// Fallback to file storage.
+	return File{}.Read(conf, serverURL)
+}
+
+// Write prefers storing the token in the operating system keyring and falls back
+// to file storage if the keyring operation fails.
+func (o KeyringWithFallback) Write(conf config.Root, serverURL *url.URL, token string) (Location, error) {
+	svc := serviceName(serverURL)
+	err := o.keyringProvider.Set(svc, keyringAccount, token)
+	if err == nil {
+		// Best effort: remove plaintext file if it exists.
+		_ = conf.Session().Delete()
```

Should we log this error?  Swallowing errors always makes me nervous 😆